src/timely-util/src/containers/stack.rs

```diff
@@ -217,14 +277,21 @@ impl<T: Columnation> ChunkedStack<T> {
     /// Note that the associated region is not initialized to a specific capacity
     /// because we can't generally know how much space would be required.
     pub fn with_capacity(capacity: usize) -> Self {
-        let local = Vec::with_capacity((capacity + Self::CHUNK - 1) / Self::CHUNK);
+        let chunks = (capacity + Self::CHUNK - 1) / Self::CHUNK;
+        let mut local = vec![];
+        local.resize_with(chunks, || Array::with_capacity(Self::CHUNK));
```

Comment:
@antiguru the code here was not actually reserving the size in the arrays so I'm doing that here. I have updated the `push` method to look at capacity to decide whether a new chunk is needed. This allows us to also reserve space from `SizableContainer::reserve`

Reply:
It was somewhat intentional to not preallocate the arrays. The cost of getting an array right now vs later should be the same, and getting it later opens up the opportunity to recycle allocations more eagerly.

Reply:
hm, that's surprising! Isn't the purpose of a `with_capacity` API to ensure that I can insert X items without extra allocations? I can certainly revert this and make the `reserve` function a no-op (i.e just add capacity to hold the number of chunks needed) but that seems like the wrong implementation.

Regarding cost, while there are no reallocations of the actual data here there is potentially temporal cost because work that can be done now is deferred for the future and now might be a more convenient time.

In general, with a `with_capacity` API all these considerations can be thought of at the calling site by using a capacity of `0` or the actual value depending on when they want the allocations to happen.

Reply:
I would say it's because `with_capacity` is a under defined API. We have to support it as it's part of the interface, and at the same time this implementation knows more about itself than the caller. At the same time, it might be the best we can offer.

I think it's very important that we do not preallocate: imagine we're forming a new largest batch in a spine, and constructing it takes a long time. We want to minimize the resource footprint during construction to avoid running out of resources. Additionally, differential chooses a capacity of the sum of input sizes, because it needs to assume that nothing will cancel. Often, we end with less than the sum, and we don't want to be in a position to carry unused memory with us, or needing to reallocate.

Reply:
> We want to minimize the resource footprint during construction to avoid running out of resources

> Additionally, differential chooses a capacity of the sum of input sizes

I'm still a bit confused because these two seem contradictory. If DD doesn't want preallocation to happen why call `with_capacity` with a large value? It sounds like DD needs to use `with_capacity(0)` and everything will work out fine when `with_capacity` does what it says on the tin. What am I missing?

Reply:
I reverted to the previous capacity behavior to unblock this PR but something doesn't feel right with how DD is using this API